Create an API endpoint to manage tickets (including custom fields)

Benbb96 opened the pull request:
I made a lot of changes inside this proposal for the API. I'm not sure if this does not break anything so this does need testing.

But I believe this could be cleaner and it displays custom fields in the API !

anust commented:
Thanks @Benbb96

gwasser commented:
Toward #198 for a new API using django restframework

gwasser commented:
Need some documentation and maybe some unit tests. An example of how to use it in practice would be great. Thanks for the work so far!

anust commented:
Any documentation to use the API calls?

gwasser commented:
@Benbb96 would you have some time to write up or share some documentation on how to use this API and django rest framework? I'm personally not very familiar with it. Thanks!

Benbb96 commented:
@gwasser Yes I think I'll have some tomorrow to add a section in the documentation. I'm willing to add some tests too.

Benbb96 commented:
@anust I've added an API section in the docs, can you check in the [Files changed](https://github.com/django-helpdesk/django-helpdesk/pull/1006/files) if the API is clear enough to you?

gwasser commented:
Thank you for your work on this @Benbb96! Documentation looks good to me. Needs some testing before getting merged into stable though.

aaminahassan commented:
**{"detail":"Invalid username/password."}** 

curl --location --request POST 'http://127.0.0.1:8000/api/tickets/'     --header 'Authorization: Basic YWRtaW46YWRtaW4='     --header 'Content-Type: application/json'     --data-raw '{"queue": 1, "title": "Test Ticket API", "description": "Test create ticket from API", "submitter_email": "aaminahassan@hotmail.com","priority": 4}'

anust commented:
I have created a user help:help, did curl request

url --location --request POST 'http://127.0.0.1:8000/api/tickets/'     --header 'Authorization: Basic aGVscDpoZWxwCg=='     --header 'Contentadmin:adminon/json'     --data-raw '{"queue": 1, "title": "Test Ticket API", "description": "Test create ticket from API", "submitter_email": "aaminahassan@hotmail.com","priority": 4}'
got error msg

`**{"detail":"Invalid username/password."}**
`

Benbb96 commented:
@anust, your Authorization and Content-Type headers seem to be wrong. Can you try with : 

    curl --location --request POST 'http://127.0.0.1:8000/api/tickets/' --header 'Authorization: Basic aGVscDpoZWxw' --header 'Content-Type: application/json' --data-raw '{"queue": 1, "title": "Test Ticket API", "description": "Test create ticket from API", "submitter_email": "aaminahassan@hotmail.com","priority": 4}'

anust commented:
Its working @Benbb96  thanks [
![Capture](https://user-images.githubusercontent.com/48667410/162394056-d46fcc72-fcc5-4a26-a8ac-0a4dfbdbfba7.JPG)
]

gwasser commented:
Great! I'm glad @anust got it working so easily, sounds like it's about ready to merge. Is it possible to make some unit tests so we don't accidentally break it in future merges?

Benbb96 commented:
@gwasser Yes, I hope to get some time this weekend to do this, either way I'll do it next week 😉

gwasser commented:
Thank you @Benbb96 ! Really appreciate your work on this PR.

Benbb96 commented:
I added some tests and I updated the documentation. I believe it's all good for that PR.  
Unless you think it would be better to create a new setting in order to activate or not the API?

gwasser commented:
Thank you @Benbb96 this looks great! I think it might be good to have the API as a setting that can be toggled on/off, since it could potentially be an unnecessary security hole for someone that doesn't intend to use it. Would it be easy to make it a setting or is that a lot of work adding if-then blocks everywhere?

Benbb96 commented:
@gwasser Yes I agree, it would be nice to have this setting in order to keep the default behavior with no API.  
I believe I can do it during this week.

Benbb96 commented:
Ok I believe it's all good for me in that PR. I created a setting `HELPDESK_ACTIVATE_API_ENDPOINT` which is False by default so it needs to be set to True to get access to the API (which is also restricted to staff users only).

I tried to fix some other tests too but I still have 16 out of 105 tests which don't pass (but this must be worked on another PR)

gwasser commented:
Thank you @Benbb96 for this very complete PR! We'll have to figure out the tests but this looks good to me for now!

anust commented:
@Benbb96 All API's are working fine, please create one for closing the ticket.

anust commented:
PATCH API closes the tickets but does not generate an email as its response or ticket follow-up.

Benbb96 commented:
Hi @anust
In fact, the API doesn't generate emails when adding a followup to a ticket. It would be a lot of work to refactor all the logic inside the update view in `staff.py` in order to reuse it in the API.
I'll try to take a look but I cannot guarantee that I'll get time to do it.

anust commented:
What should be the closing API for tickets? PATCH? Or followup?

On Mon, 15 Aug 2022 at 9:43 PM, Benbb96 ***@***.***> wrote:

> Hi @anust <https://github.com/anust>
> In fact, the API doesn't generate emails when adding a followup to a
> ticket. It would be a lot of work to refactor all the logic inside the
> update view in staff.py in order to reuse it in the API.
> I'll try to take a look but I cannot guarantee that I'll get time to do it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/django-helpdesk/django-helpdesk/pull/1006#issuecomment-1215338775>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALTJWETPSXEQHEICYKVJDRDVZJXRJANCNFSM5SFYLX4Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Best Regards

*Aamina Hassan*
NUST,Pakistan



*"Only dead fish go with the flow**!"*            *Andy Hunt*

Benbb96 commented:
For now with the current API, it should be both. You can first use the `Followup` endpoint to write a new followup then use the `Ticket` endpoint with a PATCH request in order to change ticket status to CLOSED (and fill the resolution field as well if you want).
